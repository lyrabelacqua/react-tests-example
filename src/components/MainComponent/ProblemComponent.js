function ProblemComponent() {
    console.log('testing problems ahead!');
    return (
        <div>I will cause problems on purpose!</div>
    )
}

export default  ProblemComponent;