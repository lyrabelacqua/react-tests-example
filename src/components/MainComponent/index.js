import { useEffect } from 'react';
import ProblemComponent from './ProblemComponent';

function MainComponent({ showSmallText = false, trackData, submit }) {
    useEffect(() => { trackData(); }, []);
    return (
        <div>
            This is Main Component
            <ProblemComponent />
            {showSmallText && <small>small text</small>}
            <button onClick={submit}>Submit</button>
        </div>
    )

}


export default MainComponent;