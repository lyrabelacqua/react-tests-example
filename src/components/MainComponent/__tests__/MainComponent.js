import { fireEvent, render, screen } from '@testing-library/react';
import MainComponent from '../index';
import ProblemComponent from '../ProblemComponent';

const setup = ({ showSmallText } = {}) => {
    const trackData = jest.fn();
    const submit = jest.fn();
    render(<MainComponent showSmallText={showSmallText} trackData={trackData} submit={submit}/>)
    return { trackData, submit };
};

jest.mock('../ProblemComponent', () => ({
    __esModule: true,
    default: () => <div />
}));


beforeEach(() => { console.log('!!! start')})
beforeAll(() => { console.log('!!! START');})

describe('MainComponent', () => {
    it('shows main text', () => {
        setup();
        expect(screen.getByText('This is Main Component')).toBeVisible();
    });

    it('renders small text if the prop showSmallText is set to true', () => {
        setup({ showSmallText: true });
        expect(screen.getByText('small text')).toMatchSnapshot();
    });

    it(`doesn't render small text if the prop is set to false`, () => {
        setup({ showSmallText: false });

        expect(screen.queryByText('small text')).not.toBeInTheDocument();
    });

    it('runs trackData after the component mounted', () => {
        const { trackData } = setup();

        expect(trackData).toHaveBeenCalled();
    });

    it('runs the clicking function when the button is clicked', () => {
        // arrange
        const { submit } = setup();
        
        // act
        fireEvent.click(screen.getByText('Submit'));
        
        // assert
        expect(submit).toHaveBeenCalled();

    });
});
