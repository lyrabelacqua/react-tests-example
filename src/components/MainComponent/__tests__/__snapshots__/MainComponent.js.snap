// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainComponent renders small text if the prop showSmallText is set to true 1`] = `
<small>
  small text
</small>
`;
